Add subprocess support

patricoferris opened the pull request:
This PR is primarily for generating a discussion around how best to support spawning subprocesses in Eio. In particular for issue #126. The current implementation makes the following decisions.

 - Child I/O by default is mapped to `/dev/null`
 - There's a lot of file-descriptor mangling related to child I/O which definitely needs double-checked.

One unfortunate side-effect is that adding subprocesses bypasses all capability-style security iiuc as someone could easily `cd ..` out of the current working directory of the process. Perhaps processes should live in a separate package?

Note: if it ever lands, it would probably be good to use `io-uring-spawn`: https://www.phoronix.com/news/Linux-LPC2022-io_uring_spawn

c-cube commented:
I think processes shouldn't live in another package :-). They're a cornerstone of any unix-y IO library, and having them tightly integrated with `Flow` is really nice! 

If you worry about capability security, maybe this capability should be easy to disable (in an irreversible way)? But I personally think capabilities in OCaml are more about avoiding mistakes, than true security, since there's basically no sandboxing possible.

talex5 commented on lib_eio_linux/eio_linux.ml:
```diff
@@ -1160,11 +1160,87 @@ let net = object
+  inherit Eio.Process.process
+  method pid = pid
+  method status = 
+    match Eio_unix.run_in_systhread @@ fun () -> let v = Unix.waitpid [] pid in close (); v with
```
You might need to handle `EINTR` here.

patricoferris replied:
Handled in [91c7c4a](https://github.com/ocaml-multicore/eio/pull/330/commits/91c7c4a9299e8639d313de7eb590722fe2b19baf)

talex5 commented on lib_eio_linux/eio_linux.ml:
```diff
@@ -1160,11 +1160,87 @@ let net = object
+let pid_to_process close pid = object
+  inherit Eio.Process.process
+  method pid = pid
+  method status = 
```
We need to reap the child even if the caller doesn't call `status`. So status should probably just return a promise that gets resolved automatically when it finishes (this is also what Lwt does).

patricoferris replied:
`status` now returns a promise :))

talex5 commented on lib_eio_linux/eio_linux.ml:
```diff
@@ -1160,11 +1160,87 @@ let net = object
+let process = object
+  inherit Eio.Process.t
+
+  (* TODO: Is Sys.chdir cwd domain-safe, ? *)
```
No, and neither is forcing a promise. You'll probably need some C code for this bit.

talex5 commented on lib_eio_linux/eio_linux.ml:
```diff
@@ -1160,11 +1160,87 @@ let net = object
+      Unix.close stderr
+    in
+    let process = pid_to_process close pid in
+    let cleanup () =
```
Should kill it here I think. We know that no fiber is waiting for the result, or the switch wouldn't be getting released.

patricoferris replied:
Hopefully done in https://github.com/ocaml-multicore/eio/pull/330/commits/0127acc27b8db6dbb4ae3b29ae3546597f4e98d1 -- it tries to stop the process and catches `ESRCH` if it is already gone, would it be better to remove the hook if `status` is called instead?

patricoferris replied:
Cleaning up now stops the process 👍

talex5 replied:
Not sure about whether we want an exception or not. I'm tempted to put the `waitpid` in its own fiber inside the switch, in which case it would stop the switch from exiting (unless there was an error). Then forking a process behaves like forking a fiber, keeping the switch alive.

talex5 reviewed (commented):
> Child I/O by default is mapped to /dev/null

Might be better to make these non-optional and force the caller to decide what to do. Sending errors to /dev/null by default sounds frustrating.

> One unfortunate side-effect is that adding subprocesses bypasses all capability-style security iiuc as someone could easily cd .. out of the current working directory of the process.

This is OK; it just means that the process manager is a powerful capability and you need to audit code that gets it carefully. Normally, you'd wrap it quickly in something more limited (e.g. that runs a single trusted command, or runs it inside bubblewrap, etc).

You could take a `Path.t` for the cwd. That would allow running a sub-process in a directory that wasn't in the parent's default mount namespace (and would fail if it wasn't backed by an FD).

An API like that would also be necessary on FreeBSD in `cap_enter` mode, since (presumably) there you wouldn't have access to `chdir` (only to `fchdir`).

I suggest renaming `Process.t` to `Process.mgr`, and `Process.process` to `Process.t`.

BTW, I've been idly wondering if we can (ab)use format strings for shell commands. e.g.

```ocaml
# Process.expand "rm -- %s" "My file";;
["rm"; "--"; "My file"]
```
Might be a bit of fun to get that working! Might also confuse people, though.

anmonteiro commented on lib_eio/eio.mli:
```diff
@@ -223,6 +227,10 @@ module Stdenv : sig
   *)
 
+  val process : <process : #Process.mgr as 'a; ..> -> 'a
+  (** [process t] allows spawning subprocesses. This is a powerful capability
```
I believe the 2nd sentence is a bit unnecessary here.

anmonteiro commented on lib_eio/process.mli:
```diff
@@ -0,0 +1,51 @@
+(** A process manager capable of spawning new processes. *)
+
+val spawn : sw:Switch.t -> #mgr -> ?cwd:Fs.dir Path.t -> stdin:Flow.source -> stdout:Flow.sink -> stderr:Flow.sink -> string -> string list -> t
+(** [spawn ~sw t cmd] creates a new subprocess that is connected to the
```
Can we call this `mgr`, since `t` refers to a "subprocess"?
```suggestion
(** [spawn ~sw mgr ?cwd ~stdin ~stdout ~stderr cmd args] creates a new subprocess that is connected to the
```

patricoferris replied:
Yep, much clearer :)) Fixed in https://github.com/ocaml-multicore/eio/pull/330/commits/b02b37b3a96429b1e5d920ebc1d67803e8a31892

anmonteiro commented on lib_eio/process.mli:
```diff
@@ -0,0 +1,51 @@
+    switch [sw]. The standard input and outputs redirect to nothing by default. *)
+
+val spawn_detached : #mgr -> ?cwd:Fs.dir Path.t -> stdin:Flow.source -> stdout:Flow.sink -> stderr:Flow.sink -> string -> string list -> t
+(** [spawn_detached t cmd] is like {! spawn} but the new subprocess is not
```
```suggestion
(** [spawn_detached mgr ?cwd ~stdin ~stdout ~stderr cmd args] is like {! spawn} but the new subprocess is not
```

anmonteiro commented on lib_eio/process.mli:
```diff
@@ -0,0 +1,51 @@
+
+val spawn : sw:Switch.t -> #mgr -> ?cwd:Fs.dir Path.t -> stdin:Flow.source -> stdout:Flow.sink -> stderr:Flow.sink -> string -> string list -> t
+(** [spawn ~sw t cmd] creates a new subprocess that is connected to the
+    switch [sw]. The standard input and outputs redirect to nothing by default. *)
```
I believe this sentence is wrong. These arguments are required. 

```suggestion
    switch [sw]. *)
```

anmonteiro reviewed (commented):
A few minor comments that jumped at me while reading the PR. Thanks for your work, @patricoferris

patricoferris commented:
Thanks @anmonteiro, all seemed sensible and I added the changes in https://github.com/ocaml-multicore/eio/pull/330/commits/b02b37b3a96429b1e5d920ebc1d67803e8a31892

I think I've addressed most comments there is just the relevant C code for being able to change directories for processes in the presence of multiple domains, I'll look into this soon.

talex5 commented on lib_eio/process.mli:
```diff
@@ -0,0 +1,54 @@
+    You must provide a standard input and outputs that are backed by file descriptors and
+    [cwd] will optionally change the current working directory of the process.*)
+
+val spawn_detached : #mgr -> ?cwd:Fs.dir Path.t -> stdin:Flow.source -> stdout:Flow.sink -> stderr:Flow.sink -> string -> string list -> t
```
What's the intended use of this? Is it the same as a double-fork?

patricoferris replied:
I thought this was what was needed for this point in the original issue:

> Child processes should be attached to a switch by default to avoid leaking them, but need a way to spawn detached sub-processes too.

Perhaps I misunderstood? Looking at https://stackoverflow.com/questions/881388/what-is-the-reason-for-performing-a-double-fork-when-creating-a-daemon it looks like I have, perhaps this could be added in a follow-up PR?

talex5 commented on lib_eio/unix/eio_unix.ml:
```diff
@@ -17,6 +17,7 @@ module Private = struct
     | Get_system_clock : Eio.Time.clock Effect.t
     | Socket_of_fd : Eio.Switch.t * bool * Unix.file_descr -> socket Effect.t
     | Socketpair : Eio.Switch.t * Unix.socket_domain * Unix.socket_type * int -> (socket * socket) Effect.t
+    | Pipe : Eio.Switch.t -> (<Eio.Flow.source; Eio.Flow.close; unix_fd> * <Eio.Flow.sink; Eio.Flow.close; unix_fd>) Effect.t
```
This looks useful. I suggest putting it in a separate PR so we can review it by itself.

patricoferris replied:
Done in #350 :))

talex5 commented on lib_eio_linux/eio_linux.ml:
```diff
@@ -1162,11 +1162,90 @@ let net = object
+  method needs_close = needs_close
+  method set_needs_close b = needs_close <- b
+  method status =
+    let p, r = Promise.create () in
```
You need to create the promise (and do the wait) outside of `status`, which should just await the promise that already exists. Otherwise we'll try to reap the child once for each time the user calls `status`, whereas we actually need to do it exactly once.

patricoferris replied:
Yes! Thanks, done in https://github.com/ocaml-multicore/eio/pull/330/commits/4a37efbf581251ad8021f1252509230211ba58c0

talex5 replied:
That only creates the promise outside. It's still doing the wait once for each call. It should just be:
```ocaml
method await_exit = Promise.await exit_status
```

talex5 commented on lib_eio_linux/eio_linux.ml:
```diff
@@ -1162,11 +1162,90 @@ let net = object
+      let v = wait_for_process [] pid in 
+      if needs_close then (close (); needs_close <- false);
+      match v with
+      | _, WEXITED i -> Promise.resolve r  (Eio.Process.Exited (sig_to_host_sig i))
```
```suggestion
      | _, WEXITED i -> Promise.resolve r  (Eio.Process.Exited i)
```

talex5 commented on lib_eio_linux/eio_linux.ml:
```diff
@@ -1162,11 +1162,90 @@ let net = object
+let get_fd_or_err flow =
+  match get_fd_opt flow with
+  | Some fd -> 
+    Unix.dup ~cloexec:false (FD.to_unix `Peek fd)
```
Setting an FD to be not close-on-exec needs to be done in the child process (in C). Otherwise, another domain might fork a process at the same time and inherit it.

talex5 commented on lib_eio_linux/eio_linux.ml:
```diff
@@ -1162,11 +1162,90 @@ let net = object
+    let process = pid_to_process close pid in
+    let cleanup () =
+      (* Catch if the process is already finished when trying to stop it. *)
+      (try process#stop with Unix.Unix_error (Unix.ESRCH, _, _) -> ());
```
Using `pidfd_open(2)` would likely solve these races, and would avoid the need for a systhread to do the waitpid.

talex5 commented on lib_eio_linux/eio_stubs.c:
```diff
@@ -99,3 +99,10 @@ CAMLprim value caml_eio_getdents(value v_fd) {
+
+CAMLextern int caml_convert_signal_number(int);
+
+CAMLprim value caml_signal_to_posix_signal(value v_signo)
```
What is this for?

patricoferris replied:
I should have made that clearer in the code sorry. The problem is that the OCaml signal numbers don't match the POSIX ones (or at least the ones here: https://man7.org/linux/man-pages/man7/signal.7.html). That's why earlier commits were failing with:

```
Full test results in `/src/_build/default/lib_eio_linux/tests/_build/_tests/eio_linux'.
--
355 | Test Successful in 0.003s. 5 tests run.
356 | File "tests/process.md", line 1, characters 0-0:
357 | diff --git a/_build/default/tests/process.md b/_build/default/tests/.mdx/process.md.corrected
358 | index ade46dd..1eb6ef1 100644
359 | --- a/_build/default/tests/process.md
360 | +++ b/_build/default/tests/.mdx/process.md.corrected
361 | @@ -44,7 +44,7 @@ Stopping a subprocess works and checking the status waits and reports correctly
362 | let t = spawn ~sw "sleep" [ "sleep"; "10" ] in
363 | Process.stop t;
364 | Promise.await (Process.status t);;
365 | -- : Process.status = Eio.Process.Signaled 9
366 | +- : Process.status = Eio.Process.Signaled (-7)
367 | ```
368 |  
369 | A switch will stop a process when it is released.
370 | @@ -59,7 +59,7 @@ A switch will stop a process when it is released.
371 | in
372 | run ();
373 | Promise.await @@ Process.status (Option.get !proc);;
374 | -- : Process.status = Eio.Process.Signaled 9
375 | +- : Process.status = Eio.Process.Signaled (-7)
```

as I ran the tests using Luv (which doesn't use OCaml signal numbers). So I was converting them to be the same. Do you think there's a better approach?

talex5 replied:
I don't think there is a standard set of numbers. Even the Linux man-page you linked uses different numbers for different architectures. Seems tempting to use a polymorphic variant here instead (e.g. ``Eio.Process.Signaled `TERM``.

haesbaert replied:
I had started this when I was playing with signals:
https://github.com/haesbaert/eio/blob/signal/lib_eio/include/discover.ml
Basically exports the SIGFOO and also tries to compile a program with the non standard SIGBARS that get exported as an optional. It might make sense to point out also that the Stdlib signal calls work with both the ocaml values (the negative ones) and the "real ones".

patricoferris commented on lib_eio/eio.mli:
```diff
@@ -223,6 +227,9 @@ module Stdenv : sig
       To use this, see {!Domain_manager}.
   *)
 
+  val process : <process : #Process.mgr as 'a; ..> -> 'a
```
Yep done in https://github.com/ocaml-multicore/eio/pull/330/commits/1d0f3def3087405fb55968a0089e22c1f3221be4 although now there's a slight discrepancy with having `Eio.Process` instead of `Eio.Process_mgr` but I thought it would be better to keep `Eio.Process` like `Lwt_process`.

talex5 commented:
One way to make this PR more manageable might be to first split out a PR adding subprocess support to `Eio_linux.Low_level` (or to eio-luv), then add the cross-platform support in a later PR.

patricoferris commented:
Closing in favour of #473
